bytecomp/lambda.mli

```diff
@@ -263,20 +275,24 @@ type lambda =
   | Lswitch of lambda * lambda_switch * Location.t
 (* switch on strings, clauses are sorted by string order,
    strings are pairwise distinct *)
-  | Lstringswitch of
-      lambda * (string * lambda) list * lambda option * Location.t
+  | Lstringswitch of lambda * (string * block) list * block option * Location.t
   | Lstaticraise of int * lambda list
-  | Lstaticcatch of lambda * (int * (Ident.t * value_kind) list) * lambda
-  | Ltrywith of lambda * Ident.t * lambda
-  | Lifthenelse of lambda * lambda * lambda
+  | Lstaticcatch of lambda * (int * (Ident.t * value_kind) list) * block
+  | Ltrywith of lambda * Ident.t * block
+  | Lifthenelse of lambda * block * block * Location.t
   | Lsequence of lambda * lambda
-  | Lwhile of lambda * lambda
-  | Lfor of Ident.t * lambda * lambda * direction_flag * lambda
+  | Lwhile of lambda * block * Location.t
+  | Lfor of Ident.t * lambda * lambda * direction_flag * block * Location.t
   | Lassign of Ident.t * lambda
   | Lsend of meth_kind * lambda * lambda * lambda list * Location.t
   | Levent of lambda * lambda_event
   | Lifused of Ident.t * lambda
 
+and block = private {
```

Comment:
When the "block" is a lambda node with a toplevel `Location.t` annotation, do the two locations necessarily match?

Why not go the full way and attach the location in a generic way to all lambda nodes (renaming `block` to `lambda` and `lambda` to `lambda_desc`)?

Reply:
There's no constraint with regard to the two locations you identify.  It may be preferable for these locations in some cases to be distinct -- for example if one arm of an `Lifthenelse` was a function application surrounded by `begin` and `end`, then it would be reasonable for the location on the `block` (in the `Lifthenelse` arm) to be that of the `begin` but the location on the application (inside the `block`) to be on the `f x` or whatever it is.

There's no need at present to attach the information to all nodes---at least for the main purpose of this pull request.  I did look at this, and concluded that it seemed unnecessarily disruptive.